An improved version of subgraph matcher

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82853
* __->__ #82090

This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

SherlockNoMad committed:
An improved version of subgraph matcher

[ghstack-poisoned]

SherlockNoMad commented on test/test_fx_passes.py:
```diff
@@ -3,12 +3,15 @@
 import operator
 import logging
 
+from torch import sigmoid
```
remove

ezyang commented:
Does this completely subsume the previous matcher? If so, can we replace it?

SherlockNoMad commented:
> Does this completely subsume the previous matcher? If so, can we replace it?

Yes, functionally it does. However, it returns the Match in a different (but more generally useable) format. I will update the consumer of Match in the latter PR, and then replace the previous matcher.

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches

TODOs:
- Update replace_pattern() to use this matcher
- Fix cases with identical anchor 
- Introduce wildcard matching, such Any, OneOf
- Improve node comparer to match args and kwargs values

[ghstack-poisoned]

cccclai commented:
Currently `torch/fx/subgraph_rewriter.py` use symbolic tracing under the hood, is it possible to make it more flexible, so the subgraph_rewriter can take the graph generated from other tracer (like exir.trace)? 

The current api is:
```
def replace_pattern(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
    pattern_graph = symbolic_trace(pattern).graph
    replacement_graph = symbolic_trace(replacement).graph
```

Is it possible to make it become something like
```
def replace_pattern_with_graph(gm: GraphModule, pattern: Callable, pattern_graph: Graph, replacement: Callable, replacement_graph: Graph) -> List[Match]:
    # previous logic in replace_pattern

def replace_pattern(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
    pattern_graph = symbolic_trace(pattern).graph
    replacement_graph = symbolic_trace(replacement).graph
    replace_pattern_with_graph(gm, pattern, pattern_graph, replacement, replacement_graph)

def replace_pattern_from_exir(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
    pattern_graph = exir.trace(pattern).graph
    replacement_graph = exir.trace(replacement).graph
    replace_pattern_with_graph(gm, pattern, pattern_graph, replacement, replacement_graph)

```

SherlockNoMad commented:
> Currently `torch/fx/subgraph_rewriter.py` use symbolic tracing under the hood, is it possible to make it more flexible, so the subgraph_rewriter can take the graph generated from other tracer (like exir.trace)?
> 
> The current api is:
> 
> ```
> def replace_pattern(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
>     pattern_graph = symbolic_trace(pattern).graph
>     replacement_graph = symbolic_trace(replacement).graph
> ```
> 
> Is it possible to make it become something like
> 
> ```
> def replace_pattern_with_graph(gm: GraphModule, pattern: Callable, pattern_graph: Graph, replacement: Callable, replacement_graph: Graph) -> List[Match]:
>     # previous logic in replace_pattern
> 
> def replace_pattern(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
>     pattern_graph = symbolic_trace(pattern).graph
>     replacement_graph = symbolic_trace(replacement).graph
>     replace_pattern_with_graph(gm, pattern, pattern_graph, replacement, replacement_graph)
> 
> def replace_pattern_from_exir(gm: GraphModule, pattern: Callable, replacement: Callable) -> List[Match]:
>     pattern_graph = exir.trace(pattern).graph
>     replacement_graph = exir.trace(replacement).graph
>     replace_pattern_with_graph(gm, pattern, pattern_graph, replacement, replacement_graph)
> ```

I think the most flexible API would be 

def replace_pattern(target: Union[GraphModule, Graph, Callable], )

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches

TODOs:
- Update replace_pattern() to use this matcher
- Fix cases with identical anchor 
- Introduce wildcard matching, such Any, OneOf
- Improve node comparer to match args and kwargs values

[ghstack-poisoned]

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,158 @@
+from collections import defaultdict
```
I'm expecting, somewhere here, a high level plain English explanation of the algorithm

SherlockNoMad replied:
done.  
see match() function description.

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches

TODOs:
- Update replace_pattern() to use this matcher
- Fix cases with identical anchor 
- Introduce wildcard matching, such Any, OneOf
- Improve node comparer to match args and kwargs values

[ghstack-poisoned]

ezyang commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -1,21 +1,34 @@
 
-@compatibility(is_backward_compatible=True)
-class Match(NamedTuple):
+@compatibility(is_backward_compatible=False)
```
"did you get clearance for this" lol

SherlockNoMad replied:
No... 
but this class really shouldn't be backward compatible.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,212 @@
+        4. In the case of multiple anchors, every anchor will need to find a match using step 3.
+        In addition, the matches found between anchors need to have a common intersection node
+        in order for the match to be valid. This is implemented with backtracking. See `backtracking`
+        for more details.
```
"backtracking oh nooooo"

SherlockNoMad replied:
I know... This is the first time I need to use backtracking algorithm in real work.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,212 @@
+
+        Notice: graph traversal must be done in the reverser order because a tensor can have multiple
+        consumers, but can only have a single producer. Only with reverser order, we can we jointly
+        traverse the pattern and target graph in a deterministic path.
```
well, technically, you can traverse the consumers in the order they occur chronologically, which would also make it deterministic

SherlockNoMad replied:
Sometime, after manipulating the graph, user redo the topological sort. 
So chronological order of consumer nodes is not reliable.

ezyang replied:
But... it would still be deterministic? Unless the topsort is non deterministic... which, well, isn't that your problem

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [ ] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

jamesr66a commented on test/expect/TestFXAPIBackwardCompatibility.test_class_member_back_compat-fx_backcompat_class_members.expect:
```diff
@@ -15,5 +15,4 @@ torch.fx.proxy.Attribute ['node']
 torch.fx.proxy.Proxy ['keys']
 torch.fx.proxy.TraceError []
-torch.fx.proxy.TracerBase ['check_mutable_operations', 'create_arg', 'create_node', 'create_proxy', 'iter', 'keys', 'proxy', 'proxy_buffer_attributes', 'record_stack_traces', 'to_bool', 'trace_asserts', 'traced_func_name']
-torch.fx.subgraph_rewriter.Match ['anchor', 'nodes_map']
```
As discussed offline, this PR is making a BC-breaking change without going through the proper deprecation process and is flagrantly bypassing BC enforcement mechanisms. Please do not merge this change without changes to make it BC

SherlockNoMad replied:
Updated per request.

jamesr66a reviewed (dismissed):
Do not merge

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

jamesr66a commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+__all__ = ['SubgraphMatcher']
+
+
+@compatibility(is_backward_compatible=False)
```
You don't need this if the class name starts with `_`, since it's a private API

SherlockNoMad replied:
fixed.

jamesr66a commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+
+        return True
+
+    def match(self, graph: Graph) -> List[_InternalMatch]:
```
It's incoherent to return a private class type from a public API

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+            match_placeholder: If True, placeholder node in the pattern graph will be treated as a part of
+                the targeted pattern. If False, placeholder nodes will be used a wildcard.
+            fully_contained: If True, nodes (except placeholder and output_link_nodes ) can only be consumed
+                by nodes within th pattern.
```
the

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                If False, output node is ignored during match.
+            match_placeholder: If True, placeholder node in the pattern graph will be treated as a part of
+                the targeted pattern. If False, placeholder nodes will be used a wildcard.
+            fully_contained: If True, nodes (except placeholder and output_link_nodes ) can only be consumed
```
output_link_nodes is not defined

SherlockNoMad replied:
fixed.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                 match_output: bool = False,
+                 match_placeholder: bool = False,
+                 fully_contained: bool = True,
+                 remove_overlapping_matches: bool = True) -> None:
```
That's a lot of bool configuration flags. Can we simplify?

`fully_contained` is technically unnecessary as `fully_contained = False` can be emulated by synthetically adding every intermediate node into the output node (which when `fully_contained == True` appears to denote "this node is permitted to be referenced outside of the subgraph.)

`match_output` and `match_placeholder` seem to have arisen because you are punning output/placeholder nodes to mean can be referenced outside of subgraph / pattern variable. What about having an explicit pattern variable node instead? (It would probably be a `call_function` node but you can give it a special reserved op function.)

SherlockNoMad replied:
Removed fully_contained flag.

SherlockNoMad replied:
you are right about punning output and placeholder nodes.  I guess in most of the cases, user won't need to explicitly match placeholder and output, so they won't need to worry about setting this flag. 
It' a bit less natural to explicitly create a call_function("output") node to match with the actually output. 
But If you feel strong about this, I can fix this in the upcoming PR, where I introduced the wildcard nodes.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                              returning_nodes=self.returning_nodes.copy())
+
+@compatibility(is_backward_compatible=False)
+class SubgraphMatcher:
```
Is this matcher only sound for functional graphs?

SherlockNoMad replied:
No, it can also work with non-functional graph. 
The matcher is just implemented based on Node and edge (topological connectivity) of the Graph.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                assert len(node.users) > 0, \
+                       "SubgraphMatcher cannot be initialized with an pattern with dead code"
+
+        # TODO: assert pattern is a connected graph
```
Consider:

```
x1 = f1(a)
x2 = f2(a)
return x1, x2
```

sometimes it is profitable to fuse f1 and f2 together, as they share an input node. Is this expressible?

SherlockNoMad replied:
yes, I added a test case for this pattern.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+        else:
+            # If a node has output_node as the ONLY user, then this node is a graph sink,
+            # and should be matched against as an anchor
+            self.pattern_anchors = [n for n in output_node.all_input_nodes if len(n.users) == 1]
```
Pattern anchor initialization here doesn't look quite right. If the pattern is `fully_contained`, this looks right. But if it is not, all of the internal nodes need to be anchors too. That being said, if you dump `fully_contained == False` case as I suggested above you won't have to deal with this case, it will automatically do the right thing.

SherlockNoMad replied:
I removed the fully_contained == False  case.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                    match = copy.copy(saved_match)
+
+        # TODO: Match's anchor is set to the first of self.pattern_anchors[0] for now,
+        # need to update the sematics of this field
```
I don't understand this comment. Also, semantics

SherlockNoMad replied:
fixed.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+                match_found = self._match_nodes(pattern_anchor, node, match)
+                if match_found:
+                    # match next anchor
+                    backtracking(anchor_index + 1, match)
```
Warning: exponential police 🚨🚨🚨

(even if this is the right choice and unlikely to blow up in practice, it needs to be documented clearly)

SherlockNoMad replied:
done.

ezyang commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -0,0 +1,230 @@
+            self.pattern_anchors = [n for n in output_node.all_input_nodes if len(n.users) == 1]
+
+    def _nodes_are_equal(self, pn: Node, gn: Node) -> bool:
+        # TODO: match args and kwargs
```
seems... important...

SherlockNoMad replied:
Will do in next PR.

ezyang reviewed (approved):
ok sure why not

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

SherlockNoMad committed:
Update on "An improved version of subgraph matcher"


This new version of subgraph matcher further supports 
- optionally match with pattern's placeholder and output nodes 
- patterns with multiple outputs
- filtering out non-containing matches
- filtering out overlapping matches

TODOs:
- [x] Update replace_pattern() to use this matcher
- [x] Fix cases with identical anchor 
- [x] Introduce wildcard matching, such Any, OneOf
- [ ] Improve node comparer to match args and kwargs values

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge
